sw/device/tests/plic_sw_irq_test.c

```diff
@@ -0,0 +1,117 @@
+// Copyright lowRISC contributors.
+// Licensed under the Apache License, Version 2.0, see LICENSE for details.
+// SPDX-License-Identifier: Apache-2.0
+
+#include "sw/device/lib/base/mmio.h"
+#include "sw/device/lib/dif/dif_rv_plic.h"
+#include "sw/device/lib/irq.h"
+#include "sw/device/lib/runtime/hart.h"
+#include "sw/device/lib/runtime/log.h"
+#include "sw/device/lib/testing/test_framework/check.h"
+#include "sw/device/lib/testing/test_framework/ottf_main.h"
+#include "sw/device/lib/testing/test_framework/status.h"
+
+#include "hw/top_earlgrey/sw/autogen/top_earlgrey.h"
+#include "rv_plic_regs.h"  // Generated.
+
+static const dif_rv_plic_target_t kPlicTarget = kTopEarlgreyPlicTargetIbex0;
+
+static dif_rv_plic_t plic0;
+
+static bool external_intr_triggered = false;
+static bool software_intr_triggered = false;
+static bool software_intr_triggered_multiple = false;
+
+/**
+ * Configures all interrupts in PLIC.
+ */
+static void plic_configure_irqs(dif_rv_plic_t *plic) {
+  // Set IRQ priorities to MAX and Enable the IRQ
+  for (int i = 0; i < RV_PLIC_PARAM_NUM_SRC; ++i) {
+    CHECK_DIF_OK(dif_rv_plic_irq_set_priority(plic, i, kDifRvPlicMaxPriority));
+    CHECK_DIF_OK(
+        dif_rv_plic_irq_set_enabled(plic, i, kPlicTarget, kDifToggleEnabled));
+  }
+
+  // Set Ibex IRQ priority threshold level
+  CHECK_DIF_OK(dif_rv_plic_target_set_threshold(plic, kPlicTarget,
+                                                kDifRvPlicMinPriority));
+}
+
+/**
+ * External ISR.
+ *
+ * PLIC interrupt triggers ottf_external_isr() function. The expectation is
+ * MSIP triggers this. Inside the routine, it reads all Interrupt Pending bits
+ * and confirms only MSIP occurs.
+ */
+void ottf_external_isr(void) {
+  dif_rv_plic_irq_id_t interrupt_id;
+
+  external_intr_triggered = true;
+  LOG_FATAL("External Interrupt should not occur");
+  test_status_set(kTestStatusFailed);
+
+  // Clean up external ISRs (claim & complete)
+  CHECK_DIF_OK(dif_rv_plic_irq_claim(&plic0, kPlicTarget, &interrupt_id));
+  CHECK_DIF_OK(dif_rv_plic_irq_complete(&plic0, kPlicTarget, interrupt_id));
+}
+
+/**
+ * Software ISR
+ *
+ * Only SW ISR should occur in this test.
+ */
+void ottf_software_isr(void) {
+  LOG_INFO("SOFTWARE ISR entered");
+  if (software_intr_triggered) {
+    // Should not occur multiple times
+    // Assume this is error
+    software_intr_triggered_multiple = true;
+  }
+  software_intr_triggered = true;
+
+  // Clean up software ISRs (claim? clear?)
+  CHECK_DIF_OK(dif_rv_plic_software_irq_acknowledge(&plic0, kPlicTarget));
+  LOG_INFO("Software ISR event is cleared");
+}
+
+static void execute_test(dif_rv_plic_t *plic) {
+  // Force SW IRQ
+  //CHECK_DIF_OK(dif_rv_plic_software_irq_force(plic, kPlicTarget));
+  // Above returns kDifBadArg without reason. Run the function directly
+  ptrdiff_t msip_offset = RV_PLIC_MSIP0_REG_OFFSET + (kPlicTarget * sizeof(uint32_t));
+  mmio_region_write32(plic->base_addr, msip_offset, 1);
```

Comment:
@timothytrippel I just directly implemented the `dif_rv_plic_software_irq_force` code. Then it works. Later the test got FAULT event though.. (FAULT event I am debugging now)